crypto/s2n_certificate.c

```diff
@@ -605,3 +611,113 @@ int s2n_get_cert_der(const struct s2n_cert *cert, const uint8_t **out_cert_der,
 
     return S2N_SUCCESS;
 }
+
+static int s2n_asn1_string_free(ASN1_STRING** data)
+{
+    if (*data != NULL) {
+        ASN1_STRING_free(*data);
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_get_utf8_string_from_extension_data(const uint8_t *extension_data, uint32_t extension_len, uint8_t **out_data, uint32_t *out_len)
+{
+    POSIX_ENSURE_REF(extension_data);
+    POSIX_ENSURE_GT(extension_len, 0);
+    POSIX_ENSURE_REF(out_data);
+    POSIX_ENSURE_REF(out_len);
+
+    DEFER_CLEANUP(ASN1_STRING *asn1_str = NULL, s2n_asn1_string_free);
+    const uint8_t *asn1_str_data = extension_data;
+    /* Note that d2i_ASN1_UTF8STRING increments *der_in to the byte following the parsed data.
+     * Ref: https://www.openssl.org/docs/man1.1.0/man3/d2i_ASN1_UTF8STRING.html.
+     */
+    asn1_str = d2i_ASN1_UTF8STRING(NULL, (const unsigned char **)(void *)&asn1_str_data, extension_len);
+    POSIX_ENSURE(asn1_str != NULL, S2N_ERR_INVALID_X509_EXTENSION_TYPE);
+    int type = ASN1_STRING_type(asn1_str);
+    POSIX_ENSURE(type == V_ASN1_UTF8STRING, S2N_ERR_INVALID_X509_EXTENSION_TYPE);
+    
+    *out_len = ASN1_STRING_length(asn1_str);
+    POSIX_ENSURE_GT(*out_len, 0);
```

Comment:
You need to make sure the the returned length is `<=` than `out_len`, otherwise it won't have enough space to hold the information.

Reply:
I see, yes we could use ` out_len`  as both input and output parameter to determine the size of allocated memory.